### DISCUSSION TITLE
What is the proper way to change devices if MidiKit is already running

### DISCUSSION THREAD

**Discussion by elmcapp** [orchetect/swift-midi#102] [Q&A] [answered]
What is the proper way to change devices using MidiKit?

Lets say that I have device 1 and device 2 

MidiKit is already running with device 1 assigned:
``` toOutputs: [.name("device 1") ```

Lets say that I want to change to device 2. Do we have to stop midiManager (MidiKit) then start a new instance?

**Comment by orchetect** [comment, marked as answer]
You start the manager once (usually at app launch) and then you generally shouldn't stop it until the app quits. It will dispose of itself so stopping isn't even really necessary on app quit.

With MIDI you don't generally refer to devices, 99% of the time you just deal in endpoints (sometimes referred to as "ports"). What you have there may be named "device 1" but it's really just a single port.

If "device 1" is an endpoint that exists in the system and you've connected to it (by using `.addInputConnection` for example) then you can re-use that **managed connection**. You just need to remove "device 1" from its connected endpoints, and add the new endpoint you'd like to connect to. The same receiver handler that was set up when first creating the managed connection will continue to be used.

1. Access the managed InputConnection by the tag String that was used to create it
   `let conn = midiManager.managedInputConnections[tag]`

2. Then use `conn?.remove(outputs: [.name("device 1")])` to remove the old target.
   (Ensure you use the same criteria as when it was originally added - ie: don't originally add criteria using `.uniqueID()` and then attempt to remove it by using `.name()`)
   
   Alternatively, you can use `conn?.removeAllOutputs()` to clear all targets of the connection if that is appropriate for your use case.

4. Then add the new target using `conn?.add(outputs: [])` and supply the endpoint or its criteria.

**Reply by orchetect** [reply]
This is already covered in the [Getting Started](https://github.com/orchetect/MIDIKit/wiki/Getting-Started) page but may not be totally obvious.

**Reply by elmcapp** [reply]
Yes I overload the docs, sorry. 

So I have a custom button that changes the midi devices. I put it in very simple form please excuse the coding.
I am able to change midi device like this and it works no problem
below works. 
````
 ForEach(self.midiManager.endpoints.outputs) { device in
            TextMenuItem(title: device.name) { [self] _ in
              deviceUniqueID = device.uniqueID;
              let conn = midiManager.managedInputConnections["InputConnection1"];
              conn?.removeAllOutputs()
              conn?.add(outputs: [.uniqueID(device.uniqueID)])
              
              
            }
            .updatingStateWith { [self] in deviceUniqueID == device.uniqueID ? .on : .off }
          }
````

If I press the button that will change it to all endpoints nothing happens
below does not work
````
 TextMenuItem(title: "All Midi Devices...") { [self] _ in
            deviceUniqueID = 0;
            let conn = midiManager.managedInputConnections["InputConnection1"];
            conn?.removeAllOutputs()
            conn?.mode = .allEndpoints
            conn?.filter = .owned()
          }
````

**Reply by orchetect** [reply]
Ah ok I can see what's happening - when you set the `mode` and/or `filter` property that way, the connection isn't updating. That's actually a bug - nice catch. I'll make a new issue and merge a fix.

Issue #103

**Reply by orchetect** [reply]
@elmcapp This should now be fixed in release 0.4.9
